tensorboard/webapp/widgets/line_chart_v2/sub_view/line_chart_interactive_view.ts

```diff
@@ -0,0 +1,441 @@
+  private isCursorInside = false;
+  private readonly ngUnsubscribe = new Subject();
+
+  constructor(
+    private readonly changeDetector: ChangeDetectorRef,
+    readonly scrollStrategy: RepositionScrollStrategy
+  ) {}
+
+  ngAfterViewInit() {
+    fromEvent<MouseEvent>(this.dotsContainer.nativeElement, 'dblclick', {
+      passive: false,
+    })
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((event) => {
+        // Prevent double click from selecting text.
+        event.preventDefault();
+        this.onViewExtentReset.emit();
+        this.state = InteractionState.NONE;
+        this.changeDetector.markForCheck();
+      });
+
+    fromEvent<MouseEvent>(this.dotsContainer.nativeElement, 'mousedown', {
+      passive: false,
+    })
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((event) => {
+        event.preventDefault();
+        this.state = event.shiftKey
+          ? InteractionState.PANNING
+          : InteractionState.DRAG_ZOOMING;
+        this.dragStartCoord = {x: event.offsetX, y: event.offsetY};
+
+        if (this.state === InteractionState.DRAG_ZOOMING) {
+          this.zoomBoxInUiCoordinate = {
+            x: event.offsetX,
+            width: 0,
+            y: event.offsetY,
+            height: 0,
+          };
+        }
+
+        this.changeDetector.markForCheck();
+      });
+
+    fromEvent<MouseEvent>(this.dotsContainer.nativeElement, 'mouseup', {
+      passive: true,
+    })
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe(() => {
+        this.dragStartCoord = null;
+
+        const zoomBox = this.zoomBoxInUiCoordinate;
+        if (
+          this.state === InteractionState.DRAG_ZOOMING &&
+          zoomBox.width > 0 &&
+          zoomBox.height > 0
+        ) {
+          const xMin = this.getDataX(zoomBox.x);
+          const xMax = this.getDataX(zoomBox.x + zoomBox.width);
+          const yMin = this.getDataY(zoomBox.y + zoomBox.height);
+          const yMax = this.getDataY(zoomBox.y);
+
+          this.onViewExtentChange.emit({
+            x: [xMin, xMax],
+            y: [yMin, yMax],
+          });
+        }
+        if (this.state !== InteractionState.NONE) {
+          this.state = InteractionState.NONE;
+          this.changeDetector.markForCheck();
+        }
+      });
+
+    fromEvent<MouseEvent>(this.dotsContainer.nativeElement, 'mouseenter', {
+      passive: true,
+    })
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((event) => {
+        this.isCursorInside = true;
+        this.updateTooltip(event);
+        this.changeDetector.markForCheck();
+      });
+
+    fromEvent<MouseEvent>(this.dotsContainer.nativeElement, 'mouseleave', {
+      passive: true,
+    })
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((event) => {
+        this.dragStartCoord = null;
+        this.isCursorInside = false;
+        this.updateTooltip(event);
+        this.state = InteractionState.NONE;
```

Comment:
iiuc, this means that click+drag for panning will immediately stop panning when I move the cursor outside of the chart?

I would be fine with this initially, but would note that:
- this is a small regression from existing TB line charts, which stop panning when the mouse leaves, but resumes panning when the cursor re-enters the chart region
- my preference would be to eventually allow the mouse to move anywhere on screen while drag+panning.  I think the chart region is fairly constrained for mouse movement

Reply:
Honestly, imo, stopping the panning on leaving the rect is preferable. I've seen so many times when apps misbehave when a cursor leaves the rect while mousedown, then mouseup outside of the view, then move the cursor back in the cursor; the UI, to me, feels buggier.

Reply:
The fact that other apps have buggy implementations of panning-outside-the-rect does not seem like a good reason for us to drop support for it.  Some apps with good implementations of it exist, for example Google [Maps](https://developers.google.com/maps/documentation/embed/get-started) seems to pan properly when moving the mouse outside and back in.

I don't think panning-outside-the-rect is important enough to block this PR though, so feel free to keep as is.

Reply:
Current vz-line-chart has exactly the same bug. It is a bit more involved then just "do it right" and I would rather absolve ourself from that consideration for now.